### PR TITLE
Decouple Post Blocks Experiment from Site Editor Experiment

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
Related:  https://github.com/WordPress/gutenberg/issues/26316

This PR decouples the FSE experiment from Post Blocks.  If you have a FSE theme enabled, all blocks are registered as before.

If you don't have a FSE theme, with this new option enabled, you add to the block editor the following blocks:
- SiteLogo, SiteTagline, SiteTitle
- Query
- PostTitle, PostContent, PostAuthor, PostDate, PostExcerpt, PostFeaturedImage, PostTags
- PostComment, PostCommentAuthor, PostCommentContent, PostCommentDate, PostComments, PostCommentsCount, PostCommentsForm
- PostHierarchicalTerms
				
The sole difference in the current status of this PR is the `template part` block which is enabled only with a FSE theme.

## How has this been tested?
Manually with combinations of FSE themes and the option enabled/disabled.

### CODE REVIEW

File: lib/editor-settings.php [hunk 1]
```diff
@@ -77,6 +78,7 @@ function gutenberg_get_common_block_editor_settings() {
  */
 function gutenberg_extend_post_editor_settings( $settings ) {
 	$settings['__unstableEnableFullSiteEditingBlocks'] = gutenberg_is_fse_theme();
+	$settings['__unstableEnablePostBlocks']            = gutenberg_is_experiment_enabled( 'gutenberg-post-blocks' );
```
Comment (youknowriad) [inline lib/editor-settings.php:81]: Should we add `|| gutenberg_is_fse_theme()`.  It doesn't make sense for me to have an FSE theme without these blocks.

**Review by youknowriad — COMMENTED** [review]
I believe we need to start thinking about we organize our blocks a bit better. A unique block-library package is not going to scale properly. For me there are at least three kind of blocks:

 - Raw blocks that work in any block editor instance (the ones used by asblocks)
 - Blocks that are WP specific (widgets, latest posts...) 
 - FSE Blocks

File: lib/editor-settings.php (hunk 1 above)
Comment (ntsekouras) [reply lib/editor-settings.php:81]: We do add them if `gutenberg_is_fse_theme`... The only difference in the current status of this PR is `template part`, which is enabled only in FSE.

**Comment by ntsekouras** [comment]
>I believe we need to start thinking about we organize our blocks a bit better. 

That's a great thing to discuss, yes!

How do you think `FSE Blocks` are not WP specific?

**Comment by youknowriad** [comment]
> How do you think FSE Blocks are not WP specific?

No, these are WP specific but I wanted to separate them from other wp specific blocks (not sure this distinction makes sense though)

**Comment by bph** [comment]
I like the 3-categories of blocks
- core blocks - usable with any instance of the editor  even outside WP ( Drupal, laraberg etc) 
- WP single post blocks ( also usable in block-based theme areas) 
- FSE blocks - site-editor blocks ( includes 1+2 but has specific blocks for site title, tagline, loop, mulit-post type blocks, used in special templating context) 
This will help with cognitive load to handle all this for the user ( = content creators, DIY site owners, #nocode site-inplementors )  and get it better organized. It will make clearer the context of applying settings to one post/page or to all posts etc.

**Comment by gziolo** [comment]
There is also the Classic block that we should move out of the block library. It doesn't make sense outside of the post edit and even there might be not necessary for new sites anymore. In fact, we could move it to the Block Directory but keep it shipped with WordPress so it's easy to activate when the site needs it. It might be an interesting experiment to automatically activate the Classic block when during the processing of the post content on the server, the need for it is detected. In other cases, it would get activated as soon as a user picks it from the inserter the same way it can be done for other blocks in the Block Directory.

Going further, if it would work nicely, we could try to do the same for other blocks that aren't essential for the writing experience that you classified as: "core blocks - usable with any instance of the editor even outside WP".

**Comment by mtias** [comment]
One thing we might want to double check are the categories these are currently assigned to.

**Comment by youknowriad** [comment]
To avoid big changes in one ago maybe we can export several functions instead of one but just keep all of them in `block-library` for now, we could decide later whether we want to split the package too.

**Comment by ntsekouras** [comment]
> maybe we can export several functions instead of one

What do you mean? We have `registerCoreBlocks` and `__experimentalRegisterExperimentalCoreBlocks` now.

**Comment by youknowriad** [comment]
Something like:

```
registerCoreBocks();
registerWordPressPostEditorBlocks();
registerWordPressFullSiteEditingBocks();
registerClassicBlock();
```

But maybe as experimental APIs for now, things we can use in the other `edit-*` packages until we know more whether the split makes sense, and whether they should be separate packages.

I'm just tossing an idea though and would love a lot of opinions here.

**Comment by mtias** [comment]
Reviewing this a bit more, I don't think we should introduce another experiment nor make all the blocks available at once. We should migrate the blocks we are comfortable being always available one by one. The groupings mentioned above also don't make a lot of sense to me and I'd rather lean on categories.

**Comment by skorasaurus** [comment]
Hi, 

Thanks for trying this; I'm looking forward to use the query block in my work without FSE. 

FYI, I loaded the gutenberg plugin that was built a couple hours from https://github.com/WordPress/gutenberg/pull/27822/commits/468bed6a45fa0bdabb90649164ed962dc9697565 and I enabled the option in /wp-admin/admin.php?page=gutenberg-experiments
but none of the blocks (query, postTtile) appear in the editor; wp.blocks.getBlockTypes();  doesn't return any of those blocks.

**Comment by ntsekouras** [comment]
Hi @skorasaurus - If you want to try it, my first commit was working properly: https://github.com/WordPress/gutenberg/pull/27822/commits/7b6db89344e0b8e6939a57e46151b638f169a119

In the second commit you mentioned, I started experimenting with splitting the registration of blocks and wasn't all ready.

Other than that, by exploring this, it doesn't feel great for now TBH. I've converted it to draft for now and will catch up on it after the holidays.

**Comment by carlomanf** [comment]
I'm disappointed that this proposal has gone pear-shaped. Everyone seemed to start off in agreement that the blocks should be de-coupled from FSE, and as @ntsekouras pointed out, the first commit did implement this successfully.

Then came the doubts, one after the other:

> A unique block-library package is not going to scale properly.

> There is also the Classic block that we should move out of the block library.

> I don't think we should introduce another experiment nor make all the blocks available at once.

All are valid points in the broader scope of the project, but unless I'm missing something, I don't see how they are specifically relevant to this patch or need to obstruct it. An "experiment" is only going to be temporary so I don't see the harm in introducing another one. The blocks could still be migrated one by one out of the experiment as they each become ready.

Couldn't we just merge 7b6db89 and address those other concerns later, in new tickets?

**Comment by gziolo** [comment]
We need to rethink the way we register core blocks, as already discussed. All those blocks use `block.json` that was designed with that in mind. As @youknowriad  pointed out, two main factors impact the decision of which blocks can be registered:
- a standalone editor that registers all blocks on the client only - it’s everything that can work without WordPress
- an editor embedded inside WordPress or anything else that registers blocks also on the server and shares those definitions with the client

In the WordPress core and the Gutenberg plugin, we register all blocks on the server and offer ways to modify their definitions before sending them to the client. In addition to that, we let to define the list of disallowed blocks as part of block editor settings, but it should be rather treated as something that prevents inserting blocks with UI. 

Taking all that into account makes the final proposal way more complex. In theory, we should register on the client only those core blocks that were exposed from the server. In case where there is no traditional server involved we might want to default to the limited set of essential blocks.

**Comment by mtias** [comment]
@carlomanf the point is the experiment is unnecessary and to decouple a block from FSE we need to just move its registration to the appropriate clause, which I imply we should do one by one as we deem them ready.

**Comment by ntsekouras** [comment]
Closing in favor of https://github.com/WordPress/gutenberg/pull/31423

**Pull request closed without merging**